src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -241,19 +241,275 @@ uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md,
+        ucs_assert(entry->rkey == rkey);
+        uct_ib_md_mlx5_devx_mr_lru_entry_update(md, entry, mr);
+        *is_dummy = &entry->is_dummy;
+        return UCS_ERR_ALREADY_EXISTS;
+    } else if (mr == NULL) {
+        ucs_trace("%s: mr lru size=%d miss: %x",
+                  uct_ib_device_name(&md->super.dev),
+                  kh_size(&md->lru_rkeys.hash), rkey);
+        /* trying to reset non-exist entry */
+        kh_del(rkeys, &md->lru_rkeys.hash, iter);
+        *is_dummy = NULL;
+        return UCS_ERR_NO_ELEM;
+    }
+
+    entry = ucs_malloc(sizeof(*entry), "rkey_lru_entry");
+    if (entry == NULL) {
+        ucs_error("Cannot allocate rkey LRU entry");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    /* 1st state of an entry must be non-NULL since the mr is in use or dummy */
+    ucs_assert(mr != NULL);
+    entry->indirect_mr = mr;
+    entry->rkey        = rkey;
+    entry->is_dummy    = 1;
+    ucs_list_add_tail(&md->lru_rkeys.list, &entry->list);
+    kh_val(&md->lru_rkeys.hash, iter) = entry;
+    ucs_trace("%s: mr lru size=%d push: %x->%p", uct_ib_device_name(&md->super.dev),
+              kh_size(&md->lru_rkeys.hash), rkey, mr);
+
+    if (kh_size(&md->lru_rkeys.hash) > md->super.config.max_mr_gc_queue) {
+        uct_ib_mlx5_devx_md_mr_lru_pop(md);
+    }
+
+    if ((++md->lru_rkeys.counter % (md->super.config.max_mr_gc_queue)) == 0) {
+        /* modulo prime number to avoid resonance with mkey */
+        md->mkey_tag = (md->mkey_tag + 1) % 251;
+    }
+
+    *is_dummy = &entry->is_dummy;
+    return UCS_OK;
+}
+
+static void uct_ib_mlx5_devx_mr_lru_init(uct_ib_mlx5_md_t *md)
+{
+    ucs_list_head_init(&md->lru_rkeys.list);
+    kh_init_inplace(rkeys, &md->lru_rkeys.hash);
+    md->lru_rkeys.counter = 0;
+}
+
+static void uct_ib_mlx5_devx_mr_lru_cleanup(uct_ib_mlx5_md_t *md)
+{
+    while (kh_size(&md->lru_rkeys.hash) > 0) {
+        uct_ib_mlx5_devx_md_mr_lru_pop(md);
+    }
+
+    kh_destroy_inplace(rkeys, &md->lru_rkeys.hash);
+}
+
+static ucs_status_t uct_ib_mlx5_devx_reg_indirect_key(uct_ib_md_t *ibmd,
+                                                      uct_ib_mem_t *ib_memh)
+{
+    uct_ib_mlx5_md_t *md    = ucs_derived_of(ibmd, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    uint8_t *is_dummy       = NULL;
+    unsigned i;
+    ucs_status_t status;
+
+    if (!(md->flags & UCT_IB_MLX5_MD_FLAG_KSM)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    i = 0;
+    do {
+        status = uct_ib_mlx5_devx_reg_ksm_data_contig(md,
+                                                      &memh->mrs[UCT_IB_MR_DEFAULT],
+                                                      0, &memh->indirect_dvmr,
+                                                      &memh->super.indirect_rkey);
+        if (ucs_unlikely(status != UCS_OK)) {
+            break;
+        }
+
+        status = uct_ib_md_mlx5_devx_mr_lru_push(md, memh->super.indirect_rkey,
```

Comment:
1. remove unlikely since its slowpath
2. `do { } while (status == UCS_ERR_ALREADY_EXISTS)` - remove "i" checks
3. comment that if all keys in LRU are dummy, we will always get a new key number